### PR TITLE
Cache specific network resources

### PR DESCRIPTION
**Pull request by fmoc**
This pull request adds a local network cache to every account. Using this cache, we can store resources locally to reduce the load on the server and also implement better offline support.

By default, network jobs' target resources are not stored in the cache and we always download the latest version. Therefore, currently, one needs to enable caching on a per-job basis. This might change in the future, since we depend on the server-side caching policy anyway.

Qt supports the common headers. In fact, Qt won't cache resources with `Cache-Control: ... must-revalidate` anyway (we had to check Qt's code base for this). We may have to either implement our own caching logic outside of `QNetworkAccessManager` or adapt the server policy.

Contributes to https://github.com/owncloud/client/issues/10519.

### CODE REVIEW

File: src/libsync/account.cpp [hunk 1]
```diff
@@ -160,6 +162,12 @@ void Account::setCredentials(AbstractCredentials *cred)
 
     _am = _credentials->createAM();
 
+    const QString cacheLocation = QStringLiteral("%1/accounts/%2/network/").arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation), _uuid.toString());
```
Comment (TheOneRing) [inline src/libsync/account.cpp:165]: Can you add an example.
Is the app name included?

File: src/libsync/account.cpp [hunk 2]
```diff
@@ -160,6 +162,12 @@ void Account::setCredentials(AbstractCredentials *cred)
...
+    qCDebug(lcAccount) << "Cache location for account" << this << "set to" << cacheLocation;
+    _networkCache->setCacheDirectory(cacheLocation);
+    _networkCache->setMaximumCacheSize(10 * 1024 * 1024);
```
Comment (TheOneRing) [inline src/libsync/account.cpp:168]: can you use our literals?

File: src/libsync/account.cpp [hunk 3]
```diff
@@ -160,6 +162,12 @@ void Account::setCredentials(AbstractCredentials *cred)
...
+    _am->setCache(_networkCache);
```
Comment (TheOneRing) [inline src/libsync/account.cpp:169]: The nam takes ownershipt of the cache, so after a nam reset the cache will be null.

File: src/libsync/networkjobs.cpp [hunk 4]
```diff
@@ -346,6 +346,8 @@ const QHash<QString, qint64> &PropfindJob::sizes() const
 AvatarJob::AvatarJob(AccountPtr account, const QString &userId, int size, QObject *parent)
     : AbstractNetworkJob(account, account->url(), QStringLiteral("remote.php/dav/avatars/%1/%2.png").arg(userId, QString::number(size)), parent)
 {
+    setStoreInCache(true);
```
Comment (TheOneRing) [inline src/libsync/networkjobs.cpp:349]: Requiring two calls looks like a good source for bugs?

File: src/libsync/account.cpp (hunk 1 above)
Comment (fmoc) [reply src/libsync/account.cpp:165]: `~/.cache/ownCloud/accounts/<uuid>/network` on Linux.

File: src/libsync/account.cpp (hunk 2 above)
Comment (fmoc) [reply src/libsync/account.cpp:168]: I want to get rid of our custom setting completely. We added it yesterday just for testing.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests-oc10</strong> https://drone.owncloud.com/owncloud/client/14380/6/1

<details>


</details>

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests-ocis</strong> https://drone.owncloud.com/owncloud/client/14380/7/1

<details>


</details>

File: src/libsync/networkjobs.cpp (hunk 4 above)
Comment (fmoc) [reply src/libsync/networkjobs.cpp:349]: I don't think just a subset of all possible permutations of these flags will suit our needs. We need the flexibility, though. The future network jobs added to download the server theme and according resources need to load both explicitly from the cache only(!) as well as download resources from the network (while storing this fresh response in the cache if successful). We can implement this by creating a the job twice and changing the caching policies accordingly.

I documented that both functions are related to one another. Also, the one existing example of cached resources, user avatars, uses `PreferCache` with `true` for instance. The approach attempting downloading and using the cache in parallel mentioned above will combine `AlwaysCache` with `false` and `AlwaysNetwork` with `true`.

**Comment by michaelstingl** [comment]
@fmoc please open backend issue

File: src/libsync/abstractnetworkjob.cpp [hunk 5]
```diff
@@ -155,20 +155,30 @@ void AbstractNetworkJob::sendRequest(const QByteArray &verb,
     const QNetworkRequest &req, QIODevice *requestBody)
 {
     _verb = verb;
+
     _request = req;
+    _request.setAttribute(QNetworkRequest::CacheSaveControlAttribute, _storeInCache);
+    _request.setAttribute(QNetworkRequest::CacheLoadControlAttribute, _cacheLoadControl.value());
```
Comment (TheOneRing) [inline src/libsync/abstractnetworkjob.cpp:161]: if(_cacheLoadControl)?

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/libsync/abstractnetworkjob.h [hunk 6]
```diff
@@ -130,6 +131,23 @@ class OWNCLOUDSYNC_EXPORT AbstractNetworkJob : public QObject
...
+     */
+    void setStoreInCache(bool storeInCache);
+
+    /**
+     * Configure whether to load data from the cache.
+     * By default, all resources are always fetched from the server, the cache is not being used. Qt normally would use the PreferNetwork policy (see the Qt docs for more information).
+     * See also \ref setStoreInCache
+     * @param cacheLoadControl a policy suitable for the current job
+     */
+    void setCacheLoadControl(std::optional<QNetworkRequest::CacheLoadControl> cacheLoadControl);
```
[the remark this answers is not in the document]
Comment (fmoc) [reply src/libsync/abstractnetworkjob.h:149]: You may want to clear the state again.

[the remark this answers is not in the document]
Comment (TheOneRing) [reply src/libsync/abstractnetworkjob.h:149]: I really don't think we need to expose that.
And in which scenario would you want to unset it?

**Review by TheOneRing — APPROVED** [review]

**Pull request merged**